[1단계 - 자동차 경주 구현] 재즈(함석명) 미션 제출합니다.

seokmyungham opened the pull request:
안녕하세요!! 저는 이번 6기로 참여하게된 재즈라고 합니다 ㅎㅎ
코드를 설계하면서 궁금한 부분들에 대해 여쭤보고 의견을 듣고 싶습니다.
그외에도 헷갈리는 부분들, 미흡한 부분이 정말 많아서 이것 저것 다 꼬집어주시면 정말 감사하겠습니다 ㅠㅠ

## 1. 전체적인 검증 로직 책임을 어디에 둬야할까?

입력 값 형식을 단순 검증하는 로직과 핵심 도메인과 관련된 검증 사항을 구분 지어서 생각했는데,
단순 검증의 경우 컨트롤러에서 진행하고, 핵심 검증 사항들은 각 도메인이 책임을 지도록 진행했습니다.

우선 도메인에 단순 검증까지 책임을 부여할 경우 도메인이 과도하게 무거워지는 것 같다고 생각하고 있습니다.

그리고 단순 검증을 뷰가 아닌 컨트롤러에서 진행한 이유는
프론트와 협업을 진행하는 프로젝트일 경우를 생각해보면 사용자 데이터는 가장 먼저 컨트롤러에서 만나게 되고,
프론트단에서 검증을 하더라도 서버는 항상 데이터 무결성을 고려해야 하며
서버에서 별도의 검증 과정은 필수라고 생각하는데 이 부분에 대해 어떻게 생각하시는지 정말 궁금합니다.

## 2. 단순히 검증만을 위해 기능이 없는 모델을 만들어야 하는가?

시도 횟수를 사용자 입력으로 받는 TryCount는
비즈니스와 연관된 검증 사항이 없을 경우 지금과 같이 양의 정수인지 판단하는 형식 검증이라 생각되는데

모델 클래스를 생성하기에 아무 기능이 없는 모델이 생성될 것 같아 
단순 int형으로 반환했는데 괜찮은 설계 방향인지 궁금합니다.

## 3. 출력을 위한 getter를 사용하는 것

데이터가 외부에서 변경되는 것을 막고 도메인에서 책임을 지도록 "값보다는 의미"를 반환하는 것이 옳다고 생각하고 있는데
이번 미션을 진행하며 문자열 Name을 클래스로 포장하고 뷰에서 출력하기 위해 getter를 사용하게 되었습니다.

이러면 모델의 구조가 뷰까지 혹은 외부까지 노출되는 것 같아 이러한 부분을 지양하고 싶은데
DTO를 이용해서 정말 필요한 데이터만 전달하는 방식을 선택해야할지
출력만을 위한 getter는 괜찮은지 궁금합니다.

시간 내어 글과 코드 읽어주셔서 감사하고 리뷰 기다리겠습니다!!
감사합니다.

hoyeonyy committed:
docs(README) : 구현 기능 목록 정의

hoyeonyy committed:
feat(InputView) : inputCarName 메서드 정의

- 자동차 이름 입력받는 메서드 이름 생성

hoyeonyy committed:
feat(InputView) : inputTryCount 메서드 정의

- 시도 횟수를 입력받는 메서드 생성

hoyeonyy committed:
docs(README) : 구현 기능 목록 추가

hoyeonyy committed:
feat(Car) : 클래스 생성 및 move 메서드 추가

- 자동차의 이동 상태를 1 증가시키는 메서드입니다

hoyeonyy committed:
feat(Car) : showDistance 메서드 추가

- 자동차의 이동 상태만큼 문자열을 더해서 반환합니다.

hoyeonyy committed:
test(CarTest) : move 메서드 테스트 추가

hoyeonyy committed:
feat(Name) : 클래스 생성

hoyeonyy committed:
feat(RandomNumberGenerator) : generateRandomNumber 메서드 추가

- 1에서 9사이 난수를 생성합니다.

hoyeonyy committed:
test(RandomNumberGeneratorTest) : generateRandomNumber 메서드 테스트 추가

- 반복문으로 난수를 여러번 생성하여 난수 범위를 테스트합니다.

hoyeonyy committed:
feat(Name) : get 메서드 생성

- 출력을 위한 get 메서드 생성

hoyeonyy committed:
feat(Car) : get 메서드와 정렬을 위해 Comparable 인터페이스 상속

- 출력을 위한 get 메서드 생성
- 정렬을 위해 compareTo 메서드 오버라이딩

hoyeonyy committed:
test(CarTest) : 정렬 테스트

hoyeonyy committed:
style(InputView) : 출력문 변경

hoyeonyy committed:
feat(InputController) : 입력을 받아 race와 tryNumber 메서드 추가

- race 와 tryNumber 를 입력받아 생성

hoyeonyy committed:
feat(OutputView) : showRaceResult 메서드 생성

- 처음 실행 결과를 알려주는 출력문 생성

hoyeonyy committed:
feat(OutputView) : showRoundResult 메서드 생성

- 매 라운드 마다 결과를 출력

hoyeonyy committed:
feat(OutputView) : showWinners 메서드 생성

- 최종 우승자를 출력

hoyeonyy committed:
feat(Race) : Race 클래스 생성

hoyeonyy committed:
feat(Race) : oneGame, executeRandomMove 메서드 생성

- 모든 자동차에 대해 난수를 생성하여 4이상이 나오면 자동차를 전진시키는 과정을 반복한다.

hoyeonyy committed:
feat(Race) : get 메서드와 selectWinners 메서드 생성

- selectWinners : 정렬 후 자동차의 전진 거리를 비교하여 우승자를 선정하는 메서드 생성

hoyeonyy committed:
feat(RaceController) : process 메서드 생성

- 자동차 경주의 전반적인 흐름을 제어하는 메서드 생성

hoyeonyy committed:
test(RaceTest) : 우승자가 단일, 복수일 경우를 테스트

hoyeonyy committed:
feat(Application) : 콘솔에서 실행 할수 있는 메서드 생성

hoyeonyy committed:
docs(README) : 구현 목록 동기화

hoyeonyy committed:
docs(README) : 구현 목록 추가

hoyeonyy committed:
feat(Name) : validateNameLength 메서드 생성

- 이름이 공백이거나 5글자 보다 큰 경우 예외 발생

hoyeonyy committed:
test(NameTest) : 이름이 공백이거나 5글자 초과인 경우 예외 발생 테스트

hoyeonyy committed:
feat(Convertor) : convertStringToList 메서드 생성

- 문자열을 리스트로 바꾸어주는 메서드 생성

hoyeonyy committed:
refactor(Race) : Convertor 메서드 추가

hoyeonyy committed:
test(RaceTest) : 중복된 자동차의 이름이 있다면 예외 발생

hoyeonyy committed:
feat(Race) : validateMinimumNumberOfCars 메서드 생성

- 자동차의 수가 2대 미만이라면 예외 발생

hoyeonyy committed:
test(RaceTest) : validateMinimumNumberOfCars 메서드 테스트

- 자동차의 수가 2대 미만이라면 예외 발생

hoyeonyy committed:
feat(TryCountValidator) : validateTryCountFormat 메서드 생성

- 시도 횟수가 양의 정수가 아니라면 예외 발생

hoyeonyy committed:
test(TryCountValidatorTest) : validateTryCountFormat 메서드 테스트

- 시도 횟수가 양의 정수가 아니라면 예외 발생

hoyeonyy committed:
fix(InputView) : Scanner 메서드 변경

hoyeonyy committed:
feat(InputController) : try-catch 문 생성하여 에러 발생시 재입력

- 잘못된 값이 입력되면 그 구간만 재실행

hoyeonyy committed:
style(InputController) : 가독성을 위해 static-import

hoyeonyy committed:
style(RaceController) : 가독성 향상을 위해 메서드 분리

hoyeonyy committed:
style(Car) : 가독성 향상을 위해 변수 이름 변경

hoyeonyy committed:
style(Name) : 가독성 향상을 위해 매직 리터럴 제거

hoyeonyy committed:
style(Race) : 가독성 향상을 위해 매직 리터럴 제거

hoyeonyy committed:
style : 가독성 향상을 위해 공백 추가

hoyeonyy committed:
style : 가독성 향상을 위해 매직 리터럴 추가

hoyeonyy committed:
style(Application) : 가독성 향상을 위해 메서드 명 변경

seokmyungham commented:
안녕하세요!! 코드를 설계하면서 궁금한 부분들에 대해 여쭤보고 의견을 듣고 싶습니다.

1. 검증 로직 책임을 어디에 둬야할까?

입력 값 형식을 단순 검증하는 로직과 핵심 도메인과 관련된 검증 사항을 구분지어 생각했는데,
단순 검증의 경우 컨트롤러에서 진행하고, 핵심 검증 사항들은 각 도메인이 책임을 지도록 진행했습니다.

도메인에 단순 검증까지 책임을 부여할 경우 도메인이 과도하게 무거워지는 것 같다고 생각합니다.
그리고
단순 검증을 뷰가 아닌 컨트롤러에서 검증을 진행한 이유는
프론트엔드와 협업을 진행하는 프로젝트일 경우를 생각해보면 사용자 데이터는 가장 먼저 컨트롤러에서 만나게 되고,
프론트에서 검증을 하더라도 서버는 항상 데이터 무결성을 고려해야 하고, 서버에서 별도의 검증 과정은 필수라고 생각하는데 어떻게 생각하시는지 정말 궁금합니다.

2. 단순히 검증만을 위해 기능이 없는 모델을 만들어야 하는가?
사용자 입력 값 TryCount의 경우 양의 정수인지 검증하는 단순한 검증만을 포함하고,
모델 클래스를 생성하기에는 아무 기능이 없는 모델이 생성될 것 같아 int형으로 반환했는데 이럴 경우에 어떻게 해결하는게 더 나은 방향인지 궁금합니다.

3. 출력을 위한 getter를 사용하는 것
데이터가 외부에서 변경되는 것을 막고 도메인에서 책임을 지도록 의미를 반환하는 것이 맞다고 생각하고 있는데
이번 미션을 진행하면서 문자열 Name을 클래스로 포장하고 뷰에서 출력하기 위해 getter를 사용하게 되었습니다.

이러면 모델의 구조가 노출되는 형태가 되는 것 같고 이러한 부분을 지양하고 싶은데
DTO를 이용해서 정말 필요한 데이터만 전달하는 방식을 선택해야 할 지
출력을 위한 getter는 괜찮은지 궁금합니다.

시간 내주셔서 글과 코드 읽어주셔서 감사하고 리뷰 기다리겠습니다!!
감사합니다.

Deocksoo commented on README.md:
```diff
@@ -2,6 +2,40 @@
 자동차 경주 미션 저장소
 
-## 우아한테크코스 코드리뷰
+## 구현 기능 목록
```
👍

Deocksoo replied:
👍

Deocksoo commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,36 @@
+        System.out.println(roundResult);
+    }
+
+    public static void showWinners(Race race) {
```
메서드 이름이 드러내는 것보다 더 많은 일을 하고 있네요.
단순히 winner를 보여주는 게 아니라, winner를 뽑는 일도 안에서 같이 하고 있군요.
메서드가 하는 일을 더 명확하게 드러내는 이름을 지어주거나, 혹은 메서드에서 하고있는 또다른 일을 분리해내는 것은 어떨까요?

Deocksoo commented on src/main/java/controller/RaceController.java:
```diff
@@ -0,0 +1,32 @@
+        Race race = inputController.makeRace();
+        int tryCount = inputController.makeTryCount();
+
+        showRaceResult();
```
Race 결과를 보여줄 것 같이 생겼는데, 별다른 일을 하지 않고 있네요. 꼭 필요한 메서드인가요?

Deocksoo commented on src/main/java/model/Race.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    public void oneGame() {
```
[Java naming convention](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html)에 따라 메서드가 뭘 하는지 표현할 수 있도록 조금 더 좋은 이름을 찾아주면 어떨까요~?

Deocksoo commented on src/main/java/model/Race.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private static void executeRandomMove(Car car, int randomNumber) {
+        if (randomNumber >= 4) {
```
이 숫자는 이번 게임에서 가장 중요한 규칙을 담고 있는 숫자일 것 같아요. 앞서의 리뷰와 마찬가지로 constant 처리해서 좋은 이름을 붙여주면 어떤가요?

더불어 이 부분은 아마도 자동차 미션에서 가장 핵심적인 부분 중 하나일 것 같아요.
제대로 동작한다는 것을 검증할 수 있는 테스트를 추가해보는 것은 어떤가요?

Deocksoo commented on src/main/java/model/Race.java:
```diff
@@ -0,0 +1,72 @@
+        List<String> invalidNames = convertStringToList(inValidNameStr);
+        validate(invalidNames);
+
+        for (String name : invalidNames) {
```
name은 invalidNames 에서 뽑아온 이름들이군요.
이 name은 valid한 name인가요, 아니면 invalid한 name인가요?
이 name들은 항상 invalid한 이름만 들어오나요?
이 이름으로 Car를 만들어도 괜찮나요~?

Deocksoo commented on src/main/java/model/Race.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private void validateMinimumNumberOfCars(List<String> invalidNames) {
+        if (invalidNames.size() < 2) {
```
이 숫자도 게임 규칙과 관련된 특별한 숫자일 것 같네요.
[Java naming convention](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html) 에 맞게 constant 처리하는 것은 어떤가요?

Deocksoo commented on src/main/java/model/Name.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    private void validateNameLength(String name) {
+        if (name.isBlank() || name.length() > 5) {
```
이 숫자는 게임 규칙과 관련해서 특별한 의미를 가지고 있는 숫자일 것 같아요.
[Java naming convention](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html)에 따라 constant 처리하는 것은 어떤가요?

Deocksoo commented on src/main/java/utils/TryCountValidator.java:
```diff
@@ -0,0 +1,12 @@
+
+public class TryCountValidator {
+    private static final String INVALID_FORMAT_ERROR_MESSAGE = "시도 횟수는 양의 정수여야만 합니다.";
+    private static final String format = "[1-9]\\d*";
```
Java naming convention에 따라 이름을 지어주면 좋겠어요!

Deocksoo commented on src/main/java/model/Car.java:
```diff
@@ -0,0 +1,33 @@
+        distance++;
+    }
+
+    public String showDistance() {
```
이 메서드는 distance를 보여주는 기능을 하나요?
메서드 이름과 실제 기능이 다른 것 같아요!

Deocksoo commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+public class InputView {
+    private static final String INPUT_CARS_NAME_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    private static final String INPUT_TRY_COUNT_MESSAGE = "시도할 회수는 몇회인가요?";
+    private static Scanner scanner = new Scanner(System.in);
```
Java naming convention에 따라 이름을 지어주면 좋겠어요!

Deocksoo reviewed (changes requested):
안녕하세요 재즈, 리뷰어 러너덕입니다.
작성해주신 코드 잘 읽어보았어요. 메서드를 작은 단위로 나누고, 규칙에 맞는 값이 들어오는지 잘 검증할 수 있도록 노력해주셨네요 👍 검증하기 위한 코드를 테스트하는 코드도 꼼꼼하게 잘 작성해 주셨어요.

규칙이 잘 동작하도록 꼼꼼하게 코드를 작성하고 값을 검증하는 부분은 잘 연습하고 계시는 것 같으니, 
이번 과제에서는 아래 부분들에 집중해서 코드를 작성하는 연습을 해보시면 어떨까 싶어요 :)

1. 좋은 이름 짓기
2. 중요한 규칙을 테스트하기
3. java 코드 컨벤션 익숙해지기

특히 첫 번째인 '좋은 이름 짓기'는 정말정말 중요해요. 코드를 작성하는 데에 있어서 중요한 것들이 참 많지만, 딱 한 가지를 뽑으라면 저는 '좋은 이름 짓기'를 뽑겠어요!
관련해서 고민해보시면 좋을 만한 부분들에 코멘트 남겨두었으니, 확인 한 번 부탁드려요~

ps.
> 1. 전체적인 검증 로직 책임을 어디에 둬야할까?

서버에서 필요한 검증을 하는 일은 반드시 필요하고, 중요한 일이죠!
다만 이번 미션에서는 값 검증과 관련해서 충분히 고민해주신 것 같네요 :)
이어지는 고민들은 다음 미션들을 진행하시면서 고민해보셔도 좋을 것 같아요.

> 2. 단순히 검증만을 위해 기능이 없는 모델을 만들어야 하는가?

그것을 모델이라고 부를 수 있을지는 모르겠지만, 검증하는 것도 기능일 수 있겠죠. 다만, 필요하다는 생각이 들지 않으면 만들지 않는 것이 좋아요.

> 3. 출력을 위한 getter를 사용하는 것

꼭 Dto 만들어야 하나요 ㅎㅎ 필요하면 getter 써야죠~

seokmyungham committed:
style(OutputView): 가독성 향상을 위한 메서드 명 변경

seokmyungham committed:
refactor(Car): showDistance 메서드 제거

seokmyungham committed:
feat(Car): getDistance 메서드 추가

seokmyungham committed:
refactor(Race): selectWinners 메서드 리팩토링

filter 내부에서 이동거리를 getDistance 메서드로 비교하도록 수정

seokmyungham committed:
refactor(OutputView): 모든 메서드 static 키워드 제거

seokmyungham committed:
refactor(OutputView): showWinners 메서드 리팩토링

showWinners 메서드에서 출력과 우승자 문자열을 생성하는 로직을 각각 분리

seokmyungham committed:
refactor(Race): oneGame(), executeRandomMove() 제거

seokmyungham committed:
feat(Race): start 메서드 추가

start(): 난수를 생성하고 생성한 난수가 4 이상이면 자동차를 전진시킨다

seokmyungham committed:
refactor(Car): move 메서드 리팩토링

move(): 난수를 매개변수로 받고 조건에 따라 로직이 수행되도록 수정하여 이동 조건의 책임을 자동차에게 부여

seokmyungham committed:
refactor(Race): start 메서드 리팩토링

start(): 테스트 코드 개선을 위해 조건의 책임을 메서드에서 Car 클래스의 move 메서드로 분리

seokmyungham committed:
remove(RandomNumberGenerator): 클래스 삭제

seokmyungham committed:
refactor(Car): 자동차 이동 조건을 MOVE_THRESHOLD 상수로 분리

seokmyungham committed:
style(Name): INVALID_RANGE_ERROR_MESSAGE 상수 이름 수정

seokmyungham committed:
style(InputView): 상수 명 수정

INPUT_CARS_NAME_MESSAGE -> MESSAGE_PROMPT_CAR_NAMES
INPUT_TRY_COUNT_MESSAGE -> MESSAGE_PROMPT_TRY_COUNT

seokmyungham committed:
refactor(TryCountValidator): 코드 리팩토링

정규식 문자열에서 Patten 상수를 사용하도록 코드 수정
에러 메시지 상수 이름을 TRY_COUNT_RANGE_ERROR_MESSAGE로 수정

seokmyungham committed:
style(Race): 가독성 향상을 위해 상수명 수정

seokmyungham committed:
style(TryCountValidator): 상수명을 대문자로 수정

seokmyungham committed:
style(OutputView): 메서드명 변경

seokmyungham committed:
refactor(InputView): 모든 메서드 static 키워드 제거

seokmyungham committed:
refactor(InputController): 클래스 리팩토링

seokmyungham committed:
feat(OutputView): printRaceResult, printRoundResult 메서드 리팩토링

메서드명을 변경하고 기능을 분리

seokmyungham committed:
refactor(OutputView): printResultMessage(), printRoundResult() 리팩토링

seokmyungham committed:
refactor(RaceController): 클래스 리팩토링

OutputView를 주입 받도록 변경, startRace 메서드 리팩토링

seokmyungham committed:
style(Car): 가독성 향상을 위한 개행 추가

seokmyungham committed:
refactor(Name): 상수 static 키워드 누락 수정

seokmyungham committed:
feat(InputStringParser): InputStringParser 클래스 및 메서드 이름 변경

seokmyungham committed:
refactor(Race): 리스트를 생성자 매개변수로 받도록 수정

seokmyungham committed:
refactor(InputController): getRace 메서드 내부에서 입력 문자열을 파싱하도록 변경

seokmyungham committed:
style(getTryCount): 가독성 향상을 위해 getTryCount 메서드 내부 변수명 수정

seokmyungham committed:
test: 테스트 코드 리팩토링

seokmyungham committed:
remove(RandomNumberGeneratorTest): 테스트 클래스 삭제

seokmyungham committed:
style(TryCountValidator): 가독성 향상을 위한 validateTryCountFormat 메서드 매개 변수명 수정

seokmyungham committed:
test(TryCountValidatorTest): 테스트 수정

seokmyungham committed:
test(InputStringParserTest): 테스트 추가

seokmyungham committed:
style(InputView): 가독성 향상을 위한 개행 제거

seokmyungham committed:
refactor(InputStringParser): parseInputToList 반환 타입 수정

seokmyungham committed:
refactor(Name): 가독성 향상을 위한 상수 변수 추가

seokmyungham committed:
style(Race): 상수 위치 수정

seokmyungham commented:
이렇게 페어와 구현하는 방식은 처음 경험해봤는데
즐거운 토론에 너무 열중한 나머지 마지막에 부랴부랴 기능 구현하기에도 시간이 빠듯했던 것 같아요

러너덕이 말씀한대로 충분히 고민하고 신경 써야 할 부분들에 너무 소홀했네요...
리뷰 받고 코드를 다시 보니, 정말 코드 읽기 힘드셨을 것 같아요 ㅠㅠㅠㅠ

리뷰 해주신 부분들을 다시 돌아보며 단순히 가독성을 위한 이름 짓기 이상으로,
"네이밍이 힘든 메서드는 책임이 명확하지 않거나 기능이 너무 많기 때문이다"를 몸소 경험할 수 있었어요
기능과 책임이 명확한 메서드일수록 확실히 좋은 이름을 짓기 수월함을 알게 됐습니다!!
마찬가지로 컨벤션도 의식적으로 계속 지키도록 노력하기로 다짐하게 됐어요

이름을 매우 많이 바꿔야 했었는데 조금은 좋은 이름들이 되었는지 궁금해요

> 꼭 Dto 만들어야 하나요 ㅎㅎ 필요하면 getter 써야죠~

getter를 안 써야 한다는 무언의 압박 때문에 모델의 코드가 지저분해지는 것 같았는데 
distance를 그냥 get으로 꺼낼 수 있어서 속이 시원하네요
그런데 한편으로는 모델(도메인)이 출력 뷰까지 매개변수로 들어가는 부분은 찝찝함이 조금 남아있는 것 같긴 해요
컨트롤러에서 출력을 위한 값들을 꺼내서 값만 뷰로 넘겨야 할지.. 뭔가 좀 애매한 느낌... (distance, winner 출력)

네이밍을 신경쓰다보니 자연스레 코드 구조까지 개선된 것 같아요
리뷰 해주셔서 감사드리고
코멘트 기다리고 있겠습니다!

Deocksoo reviewed (approved):
맞아요. 좋은 이름 짓기는 단순히 이름만 고민한다고 해서 되는 것이 아니라, 역할과 책임을 명확하게 나누어야 비로소 할 수 있게 되죠. 
이번 미션을 진행하면서 아주 중요한 것들을 배웠군요 :)
적절한 이름을 잘 지어주신 것 같고, 덕분에 코드도 훨씬 명확하고 깔끔하게 정리되었네요 💯

미션 진행하시느라 고생 많으셨어요. 1단계 미션은 여기서 마무리하고, 다음 미션으로 넘어가주시면 될것같아요.
